Sort signatures by keys IDs for deterministic signature order

ethan-lowman-dd opened the pull request:
In the course of implementing delegations in the repo writer ([WIP branch](https://github.com/ethan-lowman-dd/go-tuf/pull/1/files)), I had to change how keys are stored internally, which introduced non-determinism in the order of signatures. This caused tests which relied on fixtures to become flaky, as the order of signatures was not consistent. In order to keep support for static fixtures (using deterministic signature schemes like Ed25519), this PR sorts signatures by key ID.

trishankatdatadog commented on repo.go:
```diff
@@ -605,7 +607,10 @@ func (r *Repo) getSigningKeys(name string) ([]sign.Signer, error) {
-		return signingKeys, nil
+		sorted := make([]sign.Signer, len(signingKeys))
+		copy(sorted, signingKeys)
+		sort.Sort(signer.ByIDs(sorted))
```
This line is a bit hard to parse. I am sort sorting signer which are already sorted?

ethan-lowman-dd replied:
Lines 610-611 makes a copy of `signingKeys` called `sorted` (which is not yet sorted until the next line). Line 612 sorts `sorted` in place using the `signer.ByIDs` comparator.

trishankatdatadog commented on repo_test.go:
```diff
@@ -543,9 +552,17 @@ func (rs *RepoSuite) TestSign(c *C) {
+			gotKeyIDs = append(gotKeyIDs, sig.KeyID)
 		}
+
+		c.Assert(sortStrings(keyIDs), DeepEquals, sortStrings(gotKeyIDs))
```
What are we really testing here? That we sorted strings correctly? So why not just reuse the built-in function for sorting strings in the first place?

ethan-lowman-dd replied:
This tests that keyIDs and gotKeyIDs have the same elements, disregarding order. Previously the test only passed if the key IDs were in the same order.

I had a duplicate sorting of `keyIDs`/`wantKeyIDs`, which was perhaps a cause of confusion -- I've fixed this.

The `sortStrings` helper made a copy of the string slice before sorting (since `sort.Strings` sorts in place), but I removed the helper since it was confusing and only used in this one test.

rdimitrov commented:
Apart from #143, it's also worth syncing with @asraa because of #148 as both of the PRs alter/depend on the same `sign` package.

In that sense, maybe it would be better to prioritize which of these should be merged first to avoid the burden of fixing conflicts afterwards?

ethan-lowman-dd commented on internal/signer/sort.go:
```diff
@@ -0,0 +1,48 @@
+	copy(jIDs, ids)
+	sort.Strings(jIDs)
+
+	minLen := len(iIDs)
```
Go doesn't have a min function for two ints, so these four lines (33-36) are the best we can do.

asraa reviewed (commented):
Thanks! I'm not opposed to this change, but could you point me to where the testdata needed to be ordered?

I'm seeing a few checks in the interop_test/client_test but they don't seem to require ordering.

Just wondering if it makes more sense to change tests to test equality/expectation without ordering

asraa commented on repo_test.go:
```diff
@@ -543,9 +546,18 @@ func (rs *RepoSuite) TestSign(c *C) {
 		}
+		sort.Strings(gotKeyIDs)
+
+		c.Assert(wantKeyIDs, DeepEquals, gotKeyIDs)
```
client_test approaches this with https://github.com/theupdateframework/go-tuf/blob/f38e930df33905abc5bb7be5fcd0668231807a88/util/util.go#L261

ethan-lowman-dd replied:
This isn't quite the same as comparing sorted lists, right? For example:
```
a := []string{"AAA", "BBB", "CCC"}
b := []string{"AAA, "BBB", "CCC", "CCC"}

aSet := StringSliceToSet(a) 
bSet := StringSliceToSet(b)

// This evaluates to true, but the slices we're comparing are not equal.
reflect.DeepEqual(aSet, bSet)
```

I think in this case, we would want tests to fail if signatures are erroneously duplicated. Comparing sorted lists accomplishes this, but comparing sets does not.

asraa replied:
good point with duplicates. i'll have to look into that if that test makes sense to me now too

ethan-lowman-dd commented:
@asraa The tests that are flaky if signature order is nondeterministic are:

1. File hash checks against generated fixtures:
    * https://github.com/theupdateframework/go-tuf/blob/master/client/interop_test.go#L41-L48
2. Checking key IDs in order:
    * https://github.com/theupdateframework/go-tuf/blob/master/repo_test.go#L545-L548
    * https://github.com/theupdateframework/go-tuf/blob/master/repo_test.go#L1181-L1184

I addressed (1) by sorting signatures by key IDs and regenerating these fixtures. I addressed (2) by sorting keys IDs in the tests before making comparisons.

ethan-lowman-dd commented on internal/signer/sort.go:
```diff
@@ -0,0 +1,48 @@
+package signer
```
Added tests

ethan-lowman-dd commented on repo.go:
```diff
@@ -605,7 +607,10 @@ func (r *Repo) getSigningKeys(name string) ([]sign.Signer, error) {
 		return nil, err
```
With your other comment, this method no longer sorts the signers.

ethan-lowman-dd commented on repo.go:
```diff
@@ -605,7 +607,10 @@ func (r *Repo) getSigningKeys(name string) ([]sign.Signer, error) {
 		return nil, err
```
I updated the PR to sort on usage rather than on retrieval. I'm not too opinionated about which is better.

ethan-lowman-dd replied:
As discussed, I reverted this change and renamed the method to `getSortedSigningKeys`.

hosseinsia reviewed (approved):
LGTM!

hosseinsia commented on internal/signer/sort_test.go:
```diff
@@ -0,0 +1,71 @@
+
+	sort.Sort(signer.ByIDs(s))
+
+	sorted := []keys.Signer{
```
Sorry I did not realize this part last time. I think the reason for this random order is because the IDs is the hash of the value? This makes this test very unreadable. 
Can you please mock the IDs in the PublicKey to return the exact value (instead of the hash of the value)? 
and in line 59 input a random order of the signers ... in line 65 change the variable to expected and orders to be s1, s2, s3, s4, s5?

hosseinsia replied:
Hmmm I think that can not quite address the issue.
